### ISSUE TITLE
[Feature request] Multiple searches with a single request

### ISSUE THREAD

**Issue by Nick-Mazuk** [typesense/typesense#200]
## Description

In short, be able to perform multiple requests with a single HTTP call. This could be useful both searching within the same collection and searching between different collections.

Here are some examples of real-world use-cases:

- Emulating Algolia's federated search
- Allowing you to get multiple documents at once
- Probably more…

While all the use-cases are possible to do right now with multiple network requests (or some clever searches), there are some compelling reasons to allow multiple searches with a single request:

- Lower network latency
- Potentially faster search times
- Better developer experience
- Fewer sources of random errors
- Ensures data consistency between searches (i.e., there were any updates between searches, and all searches go to the same server when using the high availability feature)

Now I could see this very difficult to implement with having the entire search just be in the query parameter, so perhaps for this type of search the search data can move to the body of the request. Then you could pass multiple searches with the respective parameters into a single array.

```json
searches: [
    {
        "q": "query 1",
        "query_by": "title",
        "collection": "blogs"
    },
    {
        "q": "query 2",
        "query_by": "content",
        "collection": "comments",
        "filter": "createdAt:>124979217"
    }
]
```

And the response can be the same as it usually is, just also in an array that corresponds to each search:

```json
results: [
    {
        "facet_counts": [  ],
        "found": 1,
        "out_of": 1,
        "page": 1,
        "request_params": { "q": "query 1" },
        "search_time_ms": 1,
        "hits": [  ],
    },
    {
        "facet_counts": [  ],
        "found": 1,
        "out_of": 1,
        "page": 1,
        "request_params": { "q": "query 2" },
        "search_time_ms": 1,
        "hits": [  ],
    }
]
```

## Metadata

**Typsense Version**: 0.18 (docker)

**OS**: macOS 11.1

**Related issues**: https://github.com/typesense/typesense/issues/123, https://github.com/typesense/typesense/issues/195

**Comment by jasonbosco** [comment]
@Nick-Mazuk Multi-search is coming in `v0.19.0` :) We decided to prioritize it for the exact reasons you've outlined. 

Coincidentally, the request format and response is also exactly how you've outlined! And in addition, we're also going to support the ability to set common search parameters to be used for all searches in a batch using query params, and you can override particular search params within the body.  

Stay tuned!

**Comment by Nick-Mazuk** [comment]
Sounds great, @jasonbosco! You seem to be multiple steps ahead of me today!

**Comment by jasonbosco** [comment]
Just published v0.19.0 with this feature.
